### PR TITLE
Add determineDimensions to detect width and height

### PR DESCRIPTION
**Pull request by peter279k**
As title, it's related to issue #2. And I also implement my approach to achieve this feature :smile:.

### CODE REVIEW

File: src/Detector.php
Comment (ancarda) [inline src/Detector.php]: I really don't want to limit this to file paths on disk, as it means anyone working with a stream needs to potentially write some, or all, of the stream out to disk to use this function. Additionally, in it's current form, it's little more than a wrapper around `getimagesize`, but we can come up with a long-term solution for evolving this function:

Would you change this to accept an `SplFileInfo` and then change the `file_exists` check below to be `$stream->isReadable()`?

Anyone working with files on disk should be able to construct an `SplFileInfo` like so:

```php
$stream = new \SplFileInfo("image.jpg");
$detector->determineDimensions($stream);
```

Please don't forget to update the unit tests!

File: src/Detector.php
Comment (ancarda) [inline src/Detector.php]: As this function isn't `void` (no return value), would you please include the return type in the DocBlock, such as:

```
@return array [width, height]
```

File: src/Detector.php
Comment (ancarda) [inline src/Detector.php]: There's a typo here; "Read width **amd** height" (should be "and"?).

Also, could you please end sentences with a period (".").

File: src/Detector.php
Comment (ancarda) [inline src/Detector.php]: Please document this in the DocBlock:

```
@throws \InvalidArgumentException Unreadable SplFileInfo
```

-----

Also, as there's no test for this exception, it looks like that might actually be blocking merging because the coverage is lower. Can you see if there's any way to pass in an `SplFileInfo` that isn't readable for testing purposes?

I suppose we could try creating a PHPUnit mock that returns false for `isReadable`, and then we can assert this function was thrown.

**Review by ancarda — COMMENTED** [review]
Hi @peter279k,

Thanks so much for your pull request! This is a great start, nice work!

I've added a few notes which once applied, I should be happy to approve and merge.

**Comment by peter279k** [comment]
@ancarda, thank you for your reply.

I accept the request changes, and thank you for your review work :smile:.
If having any request changes, please feel free to write the comment to let me know :+1:.

File: src/Detector.php
Comment (ancarda) [inline src/Detector.php]: Hi @peter279k,

Would you move this outside the function? So that `determineDimensions` _accepts_ an `SplFileInfo` rather than a file path string?

Sorry if there has been some misunderstanding; I don't want the API based around filepaths, instead something more abstract as that provides more long-term capabilities.

**Review by ancarda — CHANGES_REQUESTED** [review]
Thank you again, @peter279k! Could you clear up this small, but very important detail? Please let me know if you can see reasons I can't for keeping it as a file path.

**Comment by Zegnat** [comment]
Some thoughts I have about this. Let me preface it by saying: well done on the PR! I am not sure I would have put the time in, and you did, so great kudos!

---

> Would you move this outside the function? So that `determineDimensions` accepts an `SplFileInfo` rather than a file path string?

I would make this an [`SplFileObject`](https://secure.php.net/manual/en/class.splfileobject.php). Firstly because it will match the parameter on `determineMimeType`, and secondly because I feel it makes the most sense. The [`SplFileInfo`](https://secure.php.net/manual/en/class.splfileinfo.php) class is great if you need metadata about a file from the file system itself, but isn’t going to help any of the ways an image’s dimensions can be found.

```php
public function determineDimensions(\SplFileObject $file): array
{
}
```

Now this object may or may not be a file on the file system. The easiest way to check is probably with the [`getRealPath`](https://secure.php.net/manual/en/splfileinfo.getrealpath.php) method. If it returns `false` (because the file doesn’t exist on the file system), we have to figure out a different way to get at the image dimensions. If it returns a string, just use [`getimagesize`](https://secure.php.net/manual/en/function.getimagesize.php):

```php
    if (false !== $path = $file->getRealPath()) {
        $info = getimagesize($path);
        return [$info[0], $info[1]];
    }
```

If the object doesn’t point at a local file the dimensions will need to be extracted from the accessible contents of the stream. The most straight-forward way is to use `getimagesize`’s companion function: [`getimagesizefromstring`](https://secure.php.net/manual/en/function.getimagesizefromstring.php).

Because of how file streams work the code would need to go to the start, loop through the entire stream turning it into a PHP string, and then feed it to the function:

```php
    else {
        $file->rewind();
        $buffer = '';
        while (false === $file->eof()) {
            $buffer .= $file->fread(512);
        }
        $info = getimagesizefromstring($buffer);
        return [$info[0], $info[1]];
    }
```

This mutates the original object, as the pointer will end up at the end of the stream. It’ll need code like that used by the `read` method to reset the pointer, throw proper exceptions, etc.

This has the annoying requirement of having to turn the entire (possibly multi-megabytes big) image file into an in memory string. Very much counter to how the `determineMimeType` is set-up to work with streams. The alternative is to read targetted bytes from the file only, to extract the image size. For many image formats, those things are at specific locations in the file.

Check out the interesting [FastImage](https://github.com/sdsykes/fastimage) (also [a PHP port](https://github.com/tommoor/fastimage)) for how image dimensions can be extracted from a minimal amount of file reading. After trying that there is always the possibility of falling back to `getimagesizefromstring`.

---

I hope that gets some collective synapses fired up and creativity flowing! Let me know if there is anything I can assist on.

**Comment by peter279k** [comment]
@Zegnat, thank you for your good suggestion :smile:.
To @ancarda, I think we will follow his approach to enhance the `determineDimensions` method as well :+1:.

**Comment by peter279k** [comment]
Hi @ancarda and @Zegnat , thank you for your suggestions :+1:.

Now I approve and push the latest commit about that.

Please review them and write comments if having any request changes :smiley:.

File: src/Detector.php
Comment (Zegnat) [inline src/Detector.php]: [`getimagesize`](https://secure.php.net/manual/en/function.getimagesize.php) may return `false` here (for whatever possible failure states exist). That case should probably be handled, otherwise `$info[0]` and `$info[1]` will throw.

File: src/Detector.php
Comment (Zegnat) [inline src/Detector.php]: [`SplFileObject::rewind`](https://secure.php.net/manual/en/splfileobject.rewind.php) may throw a `RuntimeException`. This should be checked and handled somehow, or it must be documented that we may throw that to the end-user. Alternatively [`SplFileObject::fseek`](https://secure.php.net/manual/en/splfileobject.fseek.php) may be used, though someone should confirm that it doesn’t throw:

```php
$file->fseek(0);
```
```

File: src/Detector.php
Comment (Zegnat) [inline src/Detector.php]: [`SplFileObject::fread`](https://secure.php.net/manual/en/splfileobject.fread.php) will mutate the object as it moves the cursor.

@ancarda is that OK? If not, before rewinding the stream this method needs to check on the cursor position and reset to it after reading. That’s what `Detector::read` does, specifically see lines 79 and 86–89:

https://github.com/ancarda/file-detector/blob/d6c9042dfb331ccd57d36a227f51689e97ac1932/src/Detector.php#L77-L91

File: src/Detector.php
Comment (Zegnat) [inline src/Detector.php]: The same as above applies to [`getimagesizefromstring`](https://secure.php.net/manual/en/function.getimagesizefromstring.php): it may return `false`.

File: tests/MimeTest.php
Comment (Zegnat) [inline tests/MimeTest.php]: This test does nothing. The Exception you are testing for is thrown by [`SplFileObject::__construct`](https://secure.php.net/manual/en/splfileobject.construct.php), not anything done by the detector.

The beauty of requiring a valid [`SplFileObject`](https://secure.php.net/manual/en/class.splfileobject.php) is that we do not need to care about this, by the time the object gets passed to `determineDimensions` we can already be sure that it is valid.

**Review by Zegnat — CHANGES_REQUESTED** [review]

File: tests/MimeTest.php
Comment (peter279k) [reply tests/MimeTest.php]: You're right. I will remove that test :+1:.

File: src/Detector.php
Comment (peter279k) [reply src/Detector.php]: I prefer adding the `RuntimeException` comment annotation to notice the end-user :+1:.

File: src/Detector.php
Comment (ancarda) [reply src/Detector.php]: Hmm, yeah, ideally we can reverse any mutations we've made so the caller doesn't need to keep track of state. That would apply for any exceptions too; we may need to catch exceptions so we can cleanup before rethrowing.

Comment (peter279k) [reply src/Detector.php]: @Zegnat, can you present the proper way about that you mention?
And please consider following code snippets:

```php
$this->read($file, 0, 512);
```

Is this correct to replace the code on line `85` with the previous code snippets?

Thanks.

Comment (Zegnat) [reply src/Detector.php]: > That would apply for any exceptions too; we may need to catch exceptions so we can cleanup before rethrowing.

Luckily neither [`SplFileObject::fread`](https://secure.php.net/manual/en/splfileobject.eof.php) nor [`SplFileObject::fread`](https://secure.php.net/manual/en/splfileobject.fread.php) throw exceptions. There should never be an unexpected exit during the copying of the stream contents to a string. And if [`SplFileObject::rewind`](https://secure.php.net/manual/en/splfileobject.rewind.php) throws that happens before any mutations (i.e. the first mutation we try fails).

> And please consider following code snippets:
>
> ```php
> $this->read($file, 0, 512);
> ```
>
> Is this correct to replace the code on line `85` with the previous code snippets?

Sadly, no, that doesn’t work. The `Detector::read` method always grabs the bytes from a specific point forward. In the case of that snippet it will start at byte 0 (start of the file) and read the first 512 bytes. You could use it by keeping track of how much has been read and updating the second argument (0) accordingly, but that seems like more work than necessary.

I also think you would incur some performance penalties calling `Detector::read` over and over again. Every time the method will do the work to reset the cursor position. That really only needs to happen once: after we’ve finished copying. (This isn’t really an issue for `Detector::determineMimeType`, which does a maximum of 2 reads.)

> can you present the proper way about that you mention?

The following code is untested, but it should give you the general idea:

```php
$position = $file->ftell();
if (0 !== $position) {
    $file->rewind();
}
$buffer = '';
while (false === $file->eof()) {
    $buffer .= $file->fread(512);
}
if (false === $position || -1 === $file->fseek($position, \SEEK_SET)) {
    throw new \Exception('Could not reset the cursor.');
}
```

Basically it does the following:

1. Store the current cursor position.
2. If the cursor isn’t at the start of the file (position 0): rewind.
3. While the cursor hasn’t reached the end of the file, read 512 bytes at a time into an in-memory string. The 512 is an arbitrary number. There might be better values. I haven’t done the research.
4. If the cursor can’t be placed back to where it was because the position is faulty or the seeking failed, throw an exception.

As I said, it is untested, so make sure to have unit tests in place if you copy any of that code. I also haven’t thought about the `Detector::read` method since I wrote it 1½ months ago, so if you see any possible enhancements (performance or otherwise), definitely contribute!

Comment (peter279k) [reply src/Detector.php]: @Zegnat, thank you for your reply. And I also follow your suggestion to implement this feature :+1: .

Please review them. If having any request changes, feel free to write the comments for that :smile: .

File: src/Detector.php
Comment (Zegnat) [inline src/Detector.php]: This change has removed the `return [$info[0], $info[1]];`. This should be put back, otherwise the code will have established the dimensions already and still continue with copying the entire file to memory.

**Review by Zegnat — CHANGES_REQUESTED** [review]

**Comment by Zegnat** [comment]
I think the code now fully covers PHP’s built in dimension functions and calling them from an SplFileObject, well done! Resolved all my previous points, just added one more request for change.

I think the final call on exception messages, the docblock, etc. should be up to @ancarda.

One thing to note: @coveralls is reporting a pretty substantial decrease in coverage. This probably needs a few more tests added!

File: src/Detector.php
Comment (peter279k) [reply src/Detector.php]: @Zegnat. Thank you for your reply. You're right and this will be put back.

The code coverage is decreased and I will add more tests to increase the code coverage!

File: tests/MimeTest.php
Comment (ancarda) [inline tests/MimeTest.php:56]: Can you add a new test at a different dimension? Just a small, single color image would work. Ideally an image that isn't square, like 3x7?

File: src/Detector.php
Comment (ancarda) [inline src/Detector.php:73]: I don't see where this exception can be thrown? Can you tell me which line(s) this exception can come from? If it's an old comment, please remove it.

**Review by ancarda — COMMENTED** [review]
@peter279k: Thanks again for all this. I'm going to leave these two as my last comments. Something will need to be done about the coverage being lower. Can you take a look at that?

File: src/Detector.php
Comment (Zegnat) [reply src/Detector.php:73]: https://github.com/ancarda/file-detector/blob/7f1a2a2dd504737e5a11a223ae44b3c5ddcefd2c/src/Detector.php#L87

See PHP documentation for [`SplFileObject::rewind`](https://secure.php.net/manual/en/splfileobject.rewind.php). Originally mentioned in [a different review comment](https://github.com/ancarda/file-detector/pull/8#discussion_r214002667).

**Comment by peter279k** [comment]
@ancarda, I've add other tests for `determineDimensions` method and try to increase the code coverage back.

Also, some uncovered codes are about the exception behaviors and it cannot let that be tested well.

File: tests/MimeTest.php
Comment (peter279k) [reply tests/MimeTest.php:56]: Ok. I will do that.

Comment (peter279k) [reply tests/MimeTest.php:56]: This test has been added via commit 558fe1.

File: src/Detector.php
Comment (peter279k) [reply src/Detector.php:73]: @ancarda. This method will throw `\InvalidArgumentException` or `\RuntimeException`.
And I think this doc block is okay.

File: tests/MimeTest.php
Comment (Zegnat) [inline tests/MimeTest.php]: Not really a problem, just wanted to point out that this line doesn’t do anything.

Comment (peter279k) [reply tests/MimeTest.php]: Hi @Zegnat, thank you for your reply.
You're right and I've removed this line :+1:.

**Comment by ancarda** [comment]
@Zegnat: If you're happy with this, I'm going to merge this.

I need to take a break from programming, likely until next year, so I'm tempted to merge this as it's essentially done (thank you @peter279k!) aside from coverage, but it's really only -1.8%, and I think 100% coverage on this project isn't possible anyway.

**Comment by Zegnat** [comment]
The feature implementation seems complete, so merging is fine with me. It is your project at the end of the day though 😉

I am happy to have a look at a separate PR for increasing coverage after merging if you want.

**Comment by ancarda** [comment]
Thanks, @Zegnat!

**Pull request merged**
